<?xml version="1.0"?>
<robot xmlns:xacro="http://wiki.ros.org/xacro" name="mia_hand">

  <!--Gazebo configuration for mia hand using DefaultRobotHWSim and mimic joint plugin-->

  <!-- hardware interface macro -->
  <xacro:macro name="transmission_macro" params="joint_name">
    <transmission name="${joint_name}_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="${joint_name}">
        <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
      <actuator name="${joint_name}_actuator">
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
    </transmission>
  </xacro:macro>

  <!--call transmission_macro macro-->
  <xacro:transmission_macro joint_name="j_index_fle" />
  <xacro:transmission_macro joint_name="j_mrl_fle" />
  <xacro:transmission_macro joint_name="j_thumb_fle" />
  <xacro:transmission_macro joint_name="j_thumb_opp" />
  <xacro:transmission_macro joint_name="j_ring_fle" /> <!--mimic joint-->
  <xacro:transmission_macro joint_name="j_little_fle" /> <!--mimic joint-->

  <!--mimic joint plugin configuration-->
  <xacro:macro name="mimic_joint_plugin_gazebo" params="name_prefix parent_joint mimic_joint has_pid:=false multiplier:=1.0 offset:=0 sensitiveness:=0.0 max_effort:=1.0 robot_namespace:=''">
    <gazebo>
      <plugin name="${name_prefix}mimic_joint_plugin" filename="libroboticsgroup_gazebo_mimic_joint_plugin.so">
        <joint>${parent_joint}</joint>
        <mimicJoint>${mimic_joint}</mimicJoint>
        <xacro:if value="${has_pid}">                     <!-- if set to true, PID parameters from "/gazebo_ros_control/pid_gains/${mimic_joint}" are loaded -->
          <hasPID />
        </xacro:if>
        <multiplier>${multiplier}</multiplier>
        <offset>${offset}</offset>
        <sensitiveness>${sensitiveness}</sensitiveness>   <!-- if absolute difference between setpoint and process value is below this threshold, do nothing; 0.0 = disable [rad] -->
        <maxEffort>${max_effort}</maxEffort>              <!-- only taken into account if has_pid:=true [Nm] -->
        <xacro:unless value="${robot_namespace == ''}">
          <robotNamespace>${robot_namespace}</robotNamespace>
        </xacro:unless>
      </plugin>
    </gazebo>
  </xacro:macro>

  <!--call mimic_joint_plugin_gazebo macro-->
  <xacro:mimic_joint_plugin_gazebo name_prefix="j_ring_fle_joint"
    parent_joint="j_mrl_fle" mimic_joint="j_ring_fle"
    has_pid="false" multiplier="1.0" max_effort="1.75" />
  <xacro:mimic_joint_plugin_gazebo name_prefix="j_little_fle_joint"
    parent_joint="j_mrl_fle" mimic_joint="j_little_fle"
    has_pid="false" multiplier="1.0" max_effort="1.75" />

  <!-- ros_control plugin to actuate the mia hand joints : its called also by universal robot arm, call it externally if needed -->
  <!--<gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so" />
  </gazebo>-->

  <!-- Gazebo Link colors for unnamed! <visual> elements,
    e.g. <visual name="foo"> will not work!
  -->
  <!--<gazebo reference="palm">--> <!-- reference to a existing "link" -->
    <!-- NOTE: use Gazebo/DepthMap to take color from mesh (.dae only) -->
    <!--<material>Gazebo/Black</material>
    <kp>1000000.0</kp>
    <kd>100.0</kd>
    <mu1>100000.0</mu1>
    <mu2>10000.0</mu2>
    <fdir1>1 0 0</fdir1>
    <selfCollide>True</selfCollide>
  </gazebo>-->

  <!-- give color and other properties to gazebo links -->
  <xacro:macro name="gazebo_properties" params="link_name color">
    <gazebo reference="${link_name}">
      <material>Gazebo/${color}</material>
      <kp>1000000.0</kp>
      <kd>100.0</kd>
      <mu1>100000.0</mu1>
      <mu2>10000.0</mu2>
      <fdir1>1 0 0</fdir1>
      <selfCollide>True</selfCollide>
    </gazebo>
  </xacro:macro>

  <!--call gazebo_properties macro-->
  <xacro:gazebo_properties link_name="palm" color="Black" />
  <xacro:gazebo_properties link_name="index_sensor" color="GreyTransparent" />
  <xacro:gazebo_properties link_name="middle_sensor" color="GreyTransparent" />
  <xacro:gazebo_properties link_name="thumb_fle" color="GreyTransparent" />
  <xacro:gazebo_properties link_name="ring_fle" color="GreyTransparent" />
  <xacro:gazebo_properties link_name="little_fle" color="GreyTransparent" />

</robot>
